qiita_ware/ebi.py

```diff
@@ -872,50 +853,68 @@ def generate_demultiplexed_fastq(self, rewrite_fastq=False, mtime=None):
 
         Notes
         -----
-        - As a performace feature, this method will check if self.ebi_dir
+        - As a performace feature, this method will check if self.full_ebi_dir
         already exists and, if it does, the script will assume that in a
         previous execution this step was performed correctly and will simply
-        read the file names from self.ebi_dir
+        read the file names from self.full_ebi_dir
         - When the object is created (init), samples, samples_prep and
         sample_demux_fps hold values for all available samples in the database.
         Here some of those values will be deleted (del's, within the loops) for
         those cases where the fastq.gz files weren't written or exist. This is
         an indication that they had no sequences and this kind of files are not
         accepted in EBI
+
+        Raises
+        ------
+        EBISubmissionError
+            - The demux file couldn't be read
+            - All samples are removed
         """
         ppd = self.preprocessed_data
 
-        if not isdir(self.ebi_dir) or rewrite_fastq:
-            makedirs(self.ebi_dir)
+        dir_not_exists = not isdir(self.full_ebi_dir)
+        if dir_not_exists or rewrite_fastq:
+            if dir_not_exists:
+                makedirs(self.full_ebi_dir)
+                filetype_id = convert_to_id('directory', 'filepath_type')
+                insert_filepaths([(self.full_ebi_dir, filetype_id)], -1,
+                                 'preprocessed_data', 'filepath',
+                                 move_files=False)
 
             demux = [path for _, path, ftype in ppd.get_filepaths()
                      if ftype == 'preprocessed_demux'][0]
 
             demux_samples = set()
-            with open_file(demux) as demux_fh:
-                for s, i in to_per_sample_ascii(demux_fh,
-                                                self.prep_template.keys()):
-                    sample_fp = self.sample_demux_fps[s]
-                    wrote_sequences = False
-                    with GzipFile(sample_fp, mode='w', mtime=mtime) as fh:
-                        for record in i:
-                            fh.write(record)
-                            wrote_sequences = True
-
-                    if wrote_sequences:
-                        demux_samples.add(s)
-                    else:
-                        del(self.samples[s])
-                        del(self.samples_prep[s])
-                        del(self.sample_demux_fps[s])
-                        remove(sample_fp)
-
+            try:
```

Comment:
Do you think it makes sense to add the traceback here? I emphasize this
because if a KeyError comes up, the error message is going to
essentially say KeyError XXX, but we won't know where that problem is
coming from, essentially obscuring the message.

On (Oct-09-15|10:38), Antonio Gonzalez wrote:
> > -                    sample_fp = self.sample_demux_fps[s]
> > -                    wrote_sequences = False
> > -                    with GzipFile(sample_fp, mode='w', mtime=mtime) as fh:
> > -                        for record in i:
> > -                            fh.write(record)
> > -                            wrote_sequences = True
> > -
> > -                    if wrote_sequences:
> > -                        demux_samples.add(s)
> > -                    else:
> > -                        del(self.samples[s])
> > -                        del(self.samples_prep[s])
> > -                        del(self.sample_demux_fps[s])
> > -                        remove(sample_fp)
> > -
> > +            try:
> 
> Good question. While writing tests I realized that one issue is that one posible issue is that the demux file doesn't exist or there are errors while reading it. Thus, that try covers any of those possible issues. I tried to isolate them but was really hard as the return is an iterator and there is no easy way to know if empty or not.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/1488/files#r41657063